verilog/analysis/flow_tree.cc

```diff
@@ -0,0 +1,348 @@
+// edges_.
+absl::Status FlowTree::GenerateControlFlowTree() {
+  // Adding edges for if blocks.
+  int current_token_enum = 0;
+  ConditionalBlock empty_block;
+  empty_block.ifdef_iterator = source_sequence_.end();
+  empty_block.ifndef_iterator = source_sequence_.end();
+  empty_block.else_iterator = source_sequence_.end();
+  empty_block.endif_iterator = source_sequence_.end();
+
+  for (TokenSequenceConstIterator iter = source_sequence_.begin();
+       iter != source_sequence_.end(); iter++) {
+    current_token_enum = iter->token_enum();
+
+    if (IsConditional(iter)) {
+      switch (current_token_enum) {
+        case PP_ifdef: {
+          if_blocks_.push_back(empty_block);
+          if_blocks_.back().ifdef_iterator = iter;
+          auto status = AddMacroOfConditionalToMap(iter);
+          if (!status.ok()) {
+            return absl::InvalidArgumentError(
+                "Error couldn't give a macro an ID.");
+          }
+          break;
+        }
+        case PP_ifndef: {
+          if_blocks_.push_back(empty_block);
+          if_blocks_.back().ifndef_iterator = iter;
+          auto status = AddMacroOfConditionalToMap(iter);
+          if (!status.ok()) {
+            return absl::InvalidArgumentError(
+                "Error couldn't give a macro an ID.");
+          }
+          break;
+        }
+        case PP_elsif: {
+          if_blocks_.back().elsif_iterators.push_back(iter);
+          auto status = AddMacroOfConditionalToMap(iter);
+          if (!status.ok()) {
+            return absl::InvalidArgumentError(
+                "Error couldn't give a macro an ID.");
+          }
+          break;
+        }
+        case PP_else: {
+          if_blocks_.back().else_iterator = iter;
+          break;
+        }
+        case PP_endif: {
+          if_blocks_.back().endif_iterator = iter;
+          auto status = AddBlockEdges(if_blocks_.back());
+          if (!status.ok()) return status;
+          // TODO(karimtera): add an error message.
+          if_blocks_.pop_back();
+          break;
+        }
+      }
+
+    } else {
+      // Only add normal edges if the next token is not `else/`elsif/`endif.
+      auto next_iter = iter + 1;
+      if (next_iter != source_sequence_.end() &&
+          next_iter->token_enum() != PP_else &&
+          next_iter->token_enum() != PP_elsif &&
+          next_iter->token_enum() != PP_endif) {
+        edges_[iter].push_back(next_iter);
+      }
+    }
+  }
+
+  return absl::OkStatus();
+}
+
+// Traveses the control flow tree in a depth first manner, appending the visited
+// tokens to current_sequence_, then provide the completed variant to the user
+// using a callback function (VariantReceiver).
+absl::Status FlowTree::DepthFirstSearch(const VariantReceiver &receiver,
+                                        TokenSequenceConstIterator current_node,
+                                        std::bitset<128> assumed) {
```

Comment:
passing the bitset by value sounds expensive.

Reply:
I can add it as a class member just like `current_macros_` bitset.
Actually, I may create a struct containing everything about the variant being generated:
1. Its `TokenSequence`.
2. bitset of visited/assumed macros so far (1 means visited or assumed, 0 means don't care or not visited).
3. bitset of assumed macros (1 means assumed to defined, 0 vice versa).

Reply:
Yes, a struct sounds good as it will help document what is needed.